# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1680983469 -7200
# Parent  58c0dfead17ac02c88874d3bcb0bc4c772b422e1
Feature: When multiple message are selected, current message is shown.

 In nsMsgDBView.cpp contains part of one hunk of Mats Palmgren's patch from
 https://bugzilla.mozilla.org/show_bug.cgi?id=364896

diff --git a/mail/base/content/folderDisplay.js b/mail/base/content/folderDisplay.js
--- a/mail/base/content/folderDisplay.js
+++ b/mail/base/content/folderDisplay.js
@@ -1453,17 +1453,32 @@ FolderDisplayWidget.prototype = {
     // content displayed mismatch, on both folder and tab changes.
     let browser = getMessagePaneBrowser();
     if (browser && browser.contentDocument && browser.contentDocument.body) {
       browser.contentDocument.body.hidden = true;
     }
 
     UpdateMailToolbar("FolderDisplayWidget displayed message changed");
     let selected = this.view.dbView.getSelectedMsgHdrs();
-    let msgHdr = selected.length ? selected[0] : null;
+    let msgHdr = null;
+    if (selected.length == 1) {
+      msgHdr = selected[0];
+    } else if (selected.length > 1) {
+      let currentIndex =  this.treeSelection?.currentIndex;
+      if (currentIndex) {
+        try {
+          msgHdr = this.view.dbView.getMsgHdrAt(currentIndex);
+        } catch(ex) {
+          msgHdr = selected[0];
+        }
+      } else {
+        // Oops, what has happened here?
+        msgHdr = selected[0];
+      }
+    }
     this.messageDisplay.onDisplayingMessage(msgHdr);
 
     // Reset gSecureMsgProbe.
     gSecureMsgProbe = {
       key: null,
       isNewRead: false,
     };
 
diff --git a/mailnews/base/src/nsMsgDBView.cpp b/mailnews/base/src/nsMsgDBView.cpp
--- a/mailnews/base/src/nsMsgDBView.cpp
+++ b/mailnews/base/src/nsMsgDBView.cpp
@@ -1220,30 +1220,29 @@ nsMsgDBView::SelectionChangedXPCOM() {
       mSummarizeFailed = true;
     }
   }
 
   bool summaryStateChanged = selectionSummarized != mSelectionSummarized;
   mSelectionSummarized = selectionSummarized;
 
   // If only one item is selected then we want to display a message.
-  if (mTreeSelection && selection.Length() == 1 && !selectionSummarized) {
-    int32_t startRange;
-    int32_t endRange;
-    nsresult rv = mTreeSelection->GetRangeAt(0, &startRange, &endRange);
+  // If we're not summarized, we display the current message.
+  if (mTreeSelection && selection.Length() >= 1 && !selectionSummarized) {
+    int32_t currentIndex;
+    nsresult rv = mTreeSelection->GetCurrentIndex(&currentIndex);
     // Tree doesn't care if we failed.
     NS_ENSURE_SUCCESS(rv, NS_OK);
 
-    if (startRange >= 0 && startRange == endRange &&
-        uint32_t(startRange) < GetSize()) {
+    if (currentIndex >= 0 && uint32_t(currentIndex) < GetSize()) {
       if (!mRemovingRow) {
         if (!mSuppressMsgDisplay)
-          LoadMessageByViewIndex(startRange);
+          LoadMessageByViewIndex(currentIndex);
         else
-          UpdateDisplayMessage(startRange);
+          UpdateDisplayMessage(currentIndex);
       }
     } else {
       // Selection seems bogus, so set to 0.
       selection.Clear();
     }
   } else {
     // If we have zero or multiple items selected, we shouldn't be displaying
     // any message.
